usr/sbin/p11sak/p11sak.c

```diff
@@ -14,3457 +14,1235 @@
-                         CK_BBOOL *forceAll, char *pqc_ver)
+static void print_help(void)
 {
-    CK_RV rc;
-    switch (cmd) {
-    case gen_key:
-        rc = generate_ckey(session, slot, kt, keylength, ECcurve, exponent,
-                label, attr_string, pqc_ver);
-        break;
-    case list_key:
-        rc = list_ckey(session, slot, kt, long_print, label, full_uri);
-        break;
-    case remove_key:
-        rc = delete_key(session, kt, label, forceAll);
-        break;
-    default:
-        fprintf(stderr, "   Unknown COMMAND %c\n", cmd);
-        print_cmd_help();
-        rc = CKR_ARGUMENTS_BAD;
-        break;
-    }
+    const struct p11sak_cmd *cmd;
 
-    return rc;
+    printf("\n");
+    printf("Usage: p11sak COMMAND [ARGS] [OPTIONS]\n");
+    printf("\n");
+    printf("COMMANDS:\n");
+    for (cmd = p11sak_commands; cmd->cmd != NULL; cmd++) {
+        printf("    %-30.30s ", cmd->cmd);
+        print_indented(cmd->description, PRINT_INDENT_POS);
+    }
+    printf("\n");
+    printf("COMMON OPTIONS\n");
+    print_options_help(p11sak_generic_opts);
+    printf("\n");
+    printf("For more information use 'p11sak COMMAND --help'.\n");
+    printf("\n");
 }
 
-static CK_RV start_session(CK_SESSION_HANDLE *session, CK_SLOT_ID slot,
-                           CK_CHAR_PTR pin, CK_ULONG pinlen)
+static void print_command_help(const struct p11sak_cmd *cmd)
 {
-    CK_SESSION_HANDLE tmp_sess;
-    CK_TOKEN_INFO tokeninfo;
-    CK_SLOT_INFO slotinfo;
-    CK_RV rc;
+    printf("\n");
+    printf("Usage: p11sak %s [ARGS] [OPTIONS]\n", cmd->cmd);
+    printf("\n");
+    printf("ARGS:\n");
+    print_arguments_help(cmd, cmd->args, 0);
+    printf("OPTIONS:\n");
+    print_options_help(cmd->opts);
+    print_options_help(p11sak_generic_opts);
+    printf("\n");
+    if (cmd->help != NULL)
+        cmd->help();
+}
 
-    rc = funcs->C_Initialize(NULL_PTR);
-    if (rc != CKR_OK) {
-        fprintf(stderr, "Error in C_Initialize (error code 0x%lX: %s)\n", rc,
-                p11_get_ckr(rc));
-        goto done;
-    }
+static void print_generate_key_attr_help(void)
+{
+    const struct p11sak_attr *attr;
 
-    rc = funcs->C_GetSlotInfo(slot, &slotinfo);
-    if (rc != CKR_OK) {
-        fprintf(stderr, "Slot %lu not available (error code 0x%lX: %s)\n", slot, rc,
-                p11_get_ckr(rc));
-        goto done;
+    printf("ATTRIBUTES:\n");
+    for (attr = p11sak_bool_attrs; attr->name != NULL; attr++) {
+        if (attr->settable)
+            printf("    '%c':   %s\n", attr->letter, attr->name);
     }
+    printf("\n");
 
-    rc = funcs->C_GetTokenInfo(slot, &tokeninfo);
-    if (rc != CKR_OK) {
-        fprintf(stderr, "Token at slot %lu not available (error code 0x%lX: %s)\n", slot,
-                rc, p11_get_ckr(rc));
-        goto done;
-    }
+    printf("    ");
+    print_indented("An uppercase letter sets the corresponding attribute to "
+                   "CK_TRUE, a lower case letter to CK_FALSE.\n"
```

Comment:
lowercase? (because of uppercase also in one word)

Reply:
fixed